### ISSUE TITLE
LLVM-COV - Warnings of "X functions have mismatched data" and missing coverage data in reports

### ISSUE THREAD

**Issue by kenkopelson** [llvm/llvm-project#72786]
This was reported back in 2017 (issue #[32849](https://github.com/llvm/llvm-project/issues/32849)), but apparently it is still an issue. After doing a fair amount of research, it seems this is a problem only with building SHARED libraries, which in this case is a DLL on Windows 10. I do not see this problem when building a STATIC library. I require using a DLL, however, so this is not something I can just ignore.

After the DLL builds, and runs all the tests, it shows the following messages on the console:

```
[build] =========================================================================================================================================
[build] All tests passed (116189 assertions in 27 test cases)
[build] 
[build] warning: 55 functions have mismatched data
[build] warning: 55 functions have mismatched data
[build] [59/59  98% :: 12.421] Install the project...

```

If it only printed those messages, I could just let this go. The bigger issue, however, is that many of the compiled files are missing from the coverage report. There is a 2-level directory structure inside of an "include" directory, a matching 2-level directory structure inside of a "src" directory. The coverage report only includes the top level of include files (with no files from the subdirectories within the include directory), and also the test files from the "test" directory.

```
|--src
    |--subdir1
    |--subdir2
|--include
    |--subdir1
    |--subdir2
|--test
```
So, in the above directory structure, only the files in include and test were included in the coverage report. When I add the -dump flag to the llvm-cov show call, I do see the 55 entries, all with 0x0 hash values.

Here is an example of the -dump output:

```
[build] warning: 55 functions have mismatched data
[build] hash-mismatch: No profile record found for '?getCode@HashCoder@@QEAAHXZ' with hash = 0x0
[build] hash-mismatch: No profile record found for '?hashCode@String@@UEBAHXZ' with hash = 0x0
[build] hash-mismatch: No profile record found for '??8String@@UEBA_NAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z' with hash = 0x0
[build] hash-mismatch: No profile record found for '??0String@@QEAA@PEBD@Z' with hash = 0x0
[build] hash-mismatch: No profile record found for '??0String@@QEAA@PEAD@Z' with hash = 0x0
[build] hash-mismatch: No profile record found for '??4String@@QEAAAEAV0@AEBV0@@Z' with hash = 0x0
[build] hash-mismatch: No profile record found for '??4String@@QEAAAEAV0@AEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z' with hash = 0x0
[build] hash-mismatch: No profile record found for '??4String@@QEAAAEAV0@PEBD@Z' with hash = 0x0
[build] hash-mismatch: No profile record found for '??BString@@QEBA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ' with hash = 0x0
[build] hash-mismatch: No profile record found for '?begin@String@@QEAA?AUiterator@1@XZ' with hash = 0x0
[build] hash-mismatch: No profile record found for '?end@String@@QEAA?AUiterator@1@XZ' with hash = 0x0
[build] hash-mismatch: No profile record found for '?length@String@@QEBA_KXZ' with hash = 0x0
[build] hash-mismatch: No profile record found for '?empty@String@@QEBA_NXZ' with hash = 0x0
[build] hash-mismatch: No profile record found for '??AString@@QEAAAEAD_K@Z' with hash = 0x0
[build] hash-mismatch: No profile record found for '??AString@@QEBAAEBD_K@Z' with hash = 0x0
```

I'm not sure what else you want me to provide. In the LLVM instructions for submitting bugs, there was no mention of this type of bug, where it really has nothing to do with the compiler.

**Comment by stevenhoving** [comment]
+1

**Comment by fatfatson** [comment]
+1 same problem, sources in dynamic library are not statisticed.

**Comment by bartlettroscoe** [comment]
I have verified that I am hitting this problem as well to with LLVM clang 20.1.1 with a shared build.  With a shared build, some of the lines of non-templated source files are not showing correct coverage.  But for a static build, they seem to be showing correct coverage.  I verified what was covered by disabling the code and rebuilding and rerunning the tests and verified that they failed (i.e. manual mutation testing).

**Comment by bearbones** [comment]
I'm encountering this with static libraries. After a good deal of investigation I found that it seems to occur only with header defined functions (ODR external linkage). The structural hash is so often `0x18` because so many of those functions are one liners or otherwise have a trivial control flow graph. Many object files include the function definition and its `profd` record, but only one TU's object file generates the "full" `covrec` record for it, marked with a `u` suffix. Others may have an empty placeholder `covrec` for it or none at all. 

From CoverageMappingGen.cpp
```
  // A dummy description for a function included-but-not-used in a TU can be
  // replaced by full description provided by a different TU. The two kinds of
  // descriptions play distinct roles: therefore, assign them different names
  // to prevent `linkonce_odr` merging.
  if (Info.IsUsed)
    FuncRecordName += "u";
```

The magic object that holds the true `covrec`, however, may not define symbols that are used in the final executable. So even if the object is in a static archive file, it can still be left out during linking. So `llvm-profdata` happily merges the `profd` records and indexes them under the correct structural hash, but `llvm-cov` finds (in the instrumented binary) only a non-`u` empty `covrec` for the function with a hash of `0x0`, and fails to find the function's profile record in `profdata` under that index. This could potentially lead to bogus region mappings too, if there is only an empty `covrec` for a function and the function's structural hash really is `0x0` per the `profd`, which I have observed in some cases.

 So I think the issue here is that the linker needs to somehow discern that the `___covrec_<MD5>u` symbol is actually "used" and so the object that contains it should be linked. 

In the meantime, my project has to force "whole archive" linking for the libraries that we want a given test binary to fully cover. But our CMake build is pretty complicated even for unit tests, and I still haven't found an elegant way to do it.

**Comment by bartlettroscoe** [comment]
> I'm encountering this with static libraries. After a good deal of investigation I found that it seems to occur only with header defined functions (ODR external linkage). 

@bearbones, are you suggesting that all template functions with implicit instantiation could have this problem if the same instantiations are generated in more than one object file?

Just to be clear, are we loosing coverage here (like in the shard lib case) or are we just getting annoying warnings about unmatched coverage data but not missing any coverage in the reports?

> In the meantime, my project has to force "whole archive" linking for the libraries that we want a given test binary to fully cover. 

You mean adding `--whole-archive <static-libs-to-link> --no-whole-archive <other-libs>` to the link lines for all tests?  Does this completely eliminate the warnings for you?

> But our CMake build is pretty complicated even for unit tests, and I still haven't found an elegant way to do it.

Does setting `-D CMAKE_CXX_LINK_LIBRARY_USING_WHOLE_ARCHIVE=ON -D CMAKE_CXX_LINK_LIBRARY_USING_WHOLE_ARCHIVE_SUPPORTED=ON` not work to set `--whole-archive` for the links of your static libs:

* https://cmake.org/cmake/help/latest/variable/CMAKE_LANG_LINK_LIBRARY_USING_FEATURE.html#predefined-features
* https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#link-features

?

I have not tired this yet (but I will after reading your post).

**Update:** But it sounds like the best solution is to fix these issue in clang itself (including resolving this issue for shared builds which would be super great).  Needing to use whole archive linking will likely explode the size of our unit test binaries.
